src/transformers/models/blt/modular_blt.py

```diff
@@ -0,0 +1,976 @@
+        Reduce variable length patches to single embedding per patch
+        Note: this works with variable number of patches for different sequences in the batch
+        It handles variable length patches by assuming that patch_lengths will be 0 for any
+        extra patches on the *right*. Since there can be a variable number of patches
+        this function also return the number of patches for each sequence in the batch.
+        Any embeddings on the right that are not allocated to a patch
+        (i.e. if the sum(patch_lengths[i]) < seq_len for any i)
+        will be sent to a dummy patch, which is trimmed before returning.
+        """
+        batch_size = hidden_states.shape[0]
+        embedding_dim = hidden_states.shape[-1]
+
+        patch_ids = patch_ids.unsqueeze(-1).expand(-1, -1, hidden_states.shape[-1])
+
+        reduced_embeddings = torch.zeros(
+            (batch_size, max_num_patches, embedding_dim), dtype=hidden_states.dtype, device=hidden_states.device
+        )
+
+        reduced_embeddings = reduced_embeddings.scatter_reduce(
+            src=hidden_states,
+            dim=1,
+            index=patch_ids,
+            reduce=reduction,
+            include_self=False,
+        )
+        reduced_embeddings = reduced_embeddings[:, :max_num_patches, :]
+
+        return reduced_embeddings
+
+
+class BltLocalDecoder(nn.Module):
+    def __init__(self, config: BltLocalDecoderConfig):
+        super().__init__()
+        self.gradient_checkpointing = False
+        self.config = config
+        self.cross_attn_decoder = True
+        self.layers = nn.ModuleList(
+            [BltTransformerLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        self.rotary_emb = BltRotaryEmbedding(config=config)
+        self.patch_embedding_projection = nn.Linear(
+            in_features=config.hidden_size_global,
+            out_features=config.hidden_size * config.cross_attn_k,
+            bias=False,
+        )
+        self.norm = BltRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.cross_attn_layers = nn.ModuleList()
+        layers_to_add = config.num_hidden_layers if config.cross_attn_all_layers else 1
+        for layer_idx in range(layers_to_add):
+            self.cross_attn_layers.append(
+                BltCrossAttention(config=config, layer_idx=layer_idx, hidden_size=config.hidden_size)
+            )
+
+    @check_model_inputs
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        patch_embeds: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Cache] = None,
+        cache_position: Optional[torch.LongTensor] = None,
+        mask: Optional[Union["BlockMask", torch.Tensor, str]] = None,
+        cross_mask: Optional[torch.Tensor] = None,
```

Comment:
Let's fix the names everywhere!!
```suggestion
        attention_mask: Optional[Union["BlockMask", torch.Tensor, str]] = None,
        cross_attention_mask: Optional[torch.Tensor] = None,
```

Reply:
updated these except when it is `encoder_attention_mask` which I updated it to in next commit :)